xdsl/interpreter.py

```diff
@@ -208,6 +209,18 @@ class Interpreter:
         default_factory=lambda: InterpreterContext(name="root")
     )
     file: IO[str] | None = field(default=None)
+    _symbol_table: dict[str, Operation] | None = None
+
+    @property
+    def symbol_table(self) -> dict[str, Operation]:
+        if self._symbol_table is None:
+            self._symbol_table = {}
+
+            for op in self.module.walk():
+                if op.has_trait(SymbolOpInterface):
+                    symbol = SymbolOpInterface.get_sym_attr_name(op)
+                    self._symbol_table[symbol.data] = op
+        return self._symbol_table
```

Comment:
How does this interact with the module being mutable?

Reply:
Only the symbol table is?
Ideally this is something shared in a SymboleTable Interface, that many operations should have.
It is an important structural dimension of MLIR we currently completely miss.
See https://mlir.llvm.org/docs/SymbolsAndSymbolTables/

Reply:
Maybe I missunderstood you question though :thinking:

Reply:
What if I query the symbol table in a rewrite, then add a symbol, then get the table again. AFAIU I will get a stale table, right?

Reply:
Oooooh, this is on the interpreter! Sorry, for some reason I thought it was on the Module. Nevermind!

Reply:
Ahah crossed mistake with mine on the other comment 😂 Seems like all is alright then 🙂